sovrn-archive/interviewMR
# JIRA-1234 calculate averageScore by NormURL hostname in MapReduce

## Calculate averageScore by NormURL hostname in MapReduce

In this PR, I wrote a MapReduce job to analyze incoming web domain "scores". The process is:
- `AverageScoreMapper`
  - use domain class `WebDomainScore` to parse tab-separated values in the format of  `src/main/data/data.tsv`
  - Finds hostname seen in NormURL (field 4) http://&lt;hostname&gt;/path
  - outputs a Text key (hostname) and FloatWritable value (score)
- `AverageScoreReducer`
  - receives an Iterator for each hostname of FloatWritable scores to average
  - calculates averageScore = totalScore / scoreCount
  - outputs hostname Text as the key, and averageScore FloatWritable as value
  - output file is tab-separated hostname, averageScore format
  - NOTE no output is produced if scoreCount is zero
## Dependency Changes
- Upgraded to JUnit 4
- Downgraded Guava for Hadoop 2.6.0 Compatibility
## How to Test this
- `mvn clean install` in root dir (with pom.xml)
- `WebDomainScoreTest` verifies handling of well-formed, malformed, or incomplete tab-separated data
- `AverageScoreMRTest` verifies the MapReduce job runs vs. sample data and for a simple example averaging 4 floats
  - You can add lines in `testSampleData()` to verify expected output: e.g. `grepOutput("runescape.wikia.com\t0.6194446");`
